### PR TITLE
[Feature] Add Traefik v2  support

### PR DESCRIPTION
**Pull request by braybaut**
## Description

This PR fixes https://github.com/alexellis/arkade/issues/12

## Motivation and Context

- [ x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Install traefik2:

`./arkade install traefik2
`

Exposing the Dashboard

`./arkade install traefik2 --dashboard=true`

Check access to traefik dashboard

`kubectl port-forward service/traefik 8181:80 --n kube-system`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [ ] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
@braybaut you DM'd several times saying that this was ready for review, but this still has no-dco and so cannot be merged. 

The note from Derek should have been of help to you, but if not, please see the contributing guide or ask for help from the community on Slack https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#paperwork-for-pull-requests

**Comment by utsavanand2** [comment]
Derek set reviewer: me

File: cmd/apps/traefik2.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		Use:          "traefik2",
+		Short:        "Install traefik2",
+		Long:         "Install traefik2",
+		Example:      `  arkade app install traefik2`,
+		SilenceUsage: true,
+	}
+
+	traefik2.Flags().StringP("namespace", "n", "kube-system", "The namespace used for installation")
+	traefik2.Flags().Bool("update-repo", true, "Update the helm repo")
+	traefik2.Flags().Bool("dashboard", false, "Expose dashboard if you want access to dashboard from the browser")
```
Comment (utsavanand2) [inline cmd/apps/traefik2.go]: @braybaut Have you tried out `traefik.Flags().BoolVar()` that way we can use `--dashboard` alone to set the value and not `--dashboard=true`. But it's totally up to you what you like to do, just a suggestion.

File: cmd/apps/traefik2.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+		"Use custom flags or override existing flags \n(example --set key=value)")
+
+	traefik2.RunE = func(command *cobra.Command, args []string) error {
+		PrintTraefikASCIIArt()
+		kubeConfigPath := getDefaultKubeconfig()
+		if command.Flags().Changed("kubeconfig") {
+			kubeConfigPath, _ = command.Flags().GetString("kubeconfig")
+		}
+		fmt.Printf("Using kubeconfig: %s\n", kubeConfigPath)
+		//arch := getNodeArchitecture()
```
Comment (utsavanand2) [inline cmd/apps/traefik2.go]: @braybaut Can we reomove these comments?

File: cmd/apps/traefik2.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+` +
+	"\n" + pkg.ThanksForUsing +
+	`
+NOTES:
+
+1. Traefik has been started. You can find out the port numbers being used by traefik by running:
+
+          $ kubectl describe svc traefik --namespace kube-system
+
+2. To Access to Dashboard you can run kubectl proxy or expose the service, for example:
```
Comment (utsavanand2) [inline cmd/apps/traefik2.go]: @braybaut Maybe we can write put the and in the next line, so the user can copy and paste the command in the terminal?

**Comment by utsavanand2** [comment]
@braybaut After making the required and suggestive changes, I would suggest you to make sure that you squash your two commits into one.
Since you have two commits right now, what you need to do is:
1. Reset the HEAD to two commits back so that the changes are there on disk but the commits are removed:
$ `git reset HEAD~2`

2. Add the files again to the staging area:
$ `git add .`

3. Commit the changes in the staging area by signing off your commit:
$ `git commit -s`
The default text editor set by git will appear where you would write your commit message.
Here is a great post to how to write great commit messages: https://chris.beams.io/posts/git-commit/

4. To avoid having merge conflicts on the upstream of the repo, I personally add another remote pointing to the upstream, for e.g., in this case do:
$ `git add remote upstream https://github.com/alexellis/arkade.git`
$ # Pull the latest commits from the upstream
$ `git pull upstream master`

5. Push your latest commit
$ `git push -f`

I know to contribute for the first time to a new OSS project can be intimidating at first, especially if you get to know so many new rules just to push a single commit. Trust me it's only hard the first time. Let me know if you need any help 😄💪

File: cmd/apps/traefik2.go (hunk 1 above)
Comment (braybaut) [reply cmd/apps/traefik2.go]: I liked this approach, I changed  to BoolVar()  and now the flag is --dashboard

**Comment by braybaut** [comment]
I did new changes that solved some points, thanks @utsavanand2  for your help.

**Comment by braybaut** [comment]
> @braybaut I tried out this PR on a GKE cluster. And everything works fine! Except I can't seem to figure out why I can't access the traefik dashboard. Do you have tested it out? If so then it would be great if you could put up a screenshot or a copy of the commands and the output. There are just a few minor changes and some suggestive changes, you can check out the comments! Thank you for this PR! Let me know if I can be of any help.

When you run `arkade install traefik2 --dashboard,`  then you just need run kubectl port-forward and access to the dashboard in your browser

Steps:

`arkade install traefik2 --dashboard
`

Exposing Dashboard:

`kubectl port-forward service/traefik 8181:80 -n kube-system
`

you can change the 8181 port  to another port that you prefer

image

File: cmd/apps/traefik2_app.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+			return err
+		}
+		fmt.Println(traefikInstallMsg)
+		return nil
+	}
+
+	return traefik2
+}
+
+func PrintTraefikASCIIArt() {
```
Comment (aidun) [inline cmd/apps/traefik2_app.go]: This could maybe a privat function.

File: cmd/apps/traefik2_app.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+		if updateRepo {
+			err = updateHelmRepos(false)
+			if err != nil {
+				return err
+			}
+		}
+
+		chartPath := path.Join(os.TempDir(), "charts")
+		err = fetchChart(chartPath, "traefik/traefik", "", false)
```
Comment (aidun) [inline cmd/apps/traefik2_app.go]: Is it possible to use a specific version or should we stay on latest?

File: cmd/apps/traefik2_app.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+		if err != nil {
+			return fmt.Errorf("Unable fetch char %s", err)
+		}
+
+		overrides := map[string]string{}
+		overrides["service.type"] = "NodePort"
```
Comment (aidun) [inline cmd/apps/traefik2_app.go]: Maybe this should be an an Option and defaults to NodePort or point out in the message how to change it.

File: cmd/apps/traefik2_app.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+func MakeInstallTraefik2() *cobra.Command {
+	var traefik2 = &cobra.Command{
+		Use:          "traefik2",
+		Short:        "Install traefik2",
+		Long:         "Install traefik2",
+		Example:      `  arkade app install traefik2`,
+		SilenceUsage: true,
+	}
+
+	var dashboard bool
```
Comment (aidun) [inline cmd/apps/traefik2_app.go]: Can you handle the `--dashbaord` flag the way you did it with `update-repo`?

**Comment by aidun** [comment]
While looking at the helm chart I saw, that Traefik is sending usage statistics by default 
```
additional:
  checkNewVersion: true
  sendAnonymousUsage: true
```
should we disable it?

File: cmd/apps/traefik2_app.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+1. Traefik has been started. You can find out the port numbers being used by traefik by running:
+
+          $ kubectl describe svc traefik --namespace kube-system
+
+2. To Access to Dashboard you can run kubectl proxy or expose the service, for example:
+
+          $ kubectl port-forward service/traefik 8181:80 -n kube-system 
+
+3. Access through https://localhost:8181/dashboard/
```
Comment (aidun) [inline cmd/apps/traefik2_app.go]: The dashboard URL is wrong. It should be `http://localhost:8181/dashboard/`

File: cmd/apps/traefik2_app.go (hunk 6 above)
Comment (alexellis) [reply cmd/apps/traefik2_app.go]: It should be a LoadBalancer normally, that's what k3s does for Traefik 1.x

**Comment by alexellis** [comment]
I think the instrumentation should be off by default too. Thank you @aidun for the detailed review

**Comment by alexellis** [comment]
What's the reason that we're not using helm3?

**Comment by alexellis** [comment]
@aidun did I miss anything in #40 ?

**Comment by alexellis** [comment]
Hi @braybaut thank you for this PR, I've taken your changes in #40 and added the feedback from @aidun 

Both, please feel free to follow-up with anything you feel is necessary in another PR.

Alex

**Comment by wingkwong** [comment]
Following @braybaut's steps , the installation was fine but it showed ``404 page not found`` when visiting ``http://localhost:<port>/dashboard/``. Any idea?

**Pull request closed without merging**